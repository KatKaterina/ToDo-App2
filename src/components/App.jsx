import React, { useState, useRef, useEffect } from 'react';
import {
  Row,
  Container,
  Col,
  Form,
} from 'react-bootstrap';

export const FormTask = ({ addTask, changeTask }) => {
  const [textTask, setTextTask] = useState('');

  const inputRef = useRef();

  useEffect(() => {
    inputRef.current.focus();
  }, []);

  const handleChange = (e) => {
    setTextTask(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    addTask(textTask);
    setTextTask('');
  };

  const handleClick = () => {
    changeTask();
  };

  return (
    <Row className="new-task">
      <div className="col-sm-2">
        <button type="button" className="button-control" onClick={handleClick}>âœ“</button>
      </div>
      <Col>
        <Form onSubmit={handleSubmit} role="form">
          <Row>
            <Col>
              <input className="input-task" type="text" data-testid="input-task" placeholder="What needs to be done?" value={textTask} onChange={handleChange} ref={inputRef} />
            </Col>
            <div className="col-sm-2">
              <button type="submit" className="button-control" onClick={handleSubmit}>+</button>
            </div>
          </Row>
        </Form>
      </Col>
    </Row>
  );
};

export const Tasks = (props) => {
  const {
    allTasks,
    filterTask,
    changeTask,
    deleteTask,
  } = props;

  let filtredTasks = allTasks;
  if (filterTask === 'active') {
    filtredTasks = allTasks.filter(({ isDone }) => !isDone);
  } else if (filterTask === 'completed') {
    filtredTasks = allTasks.filter(({ isDone }) => isDone);
  }

  const handleChange = (id) => () => {
    changeTask(id);
  };

  const handleDelete = (id) => () => {
    deleteTask(id);
  };

  return (
    <div>
      {filtredTasks.map(({ id, text, isDone }) => (
        <Row key={id} className="task align-items-center">
          <div className="col-sm-2">
            <button
              className={isDone ? 'button-control checked' : 'button-control unchecked'}
              onClick={handleChange(id)}
              data-testid="button-checked"
              type="button"
            />
          </div>
          <Col>
            <Row>
              <div className="col">
                <span className={isDone ? 'complite-task' : 'active-task'}>{text}</span>
              </div>
              <div className="col-sm-2">
                <button
                  className="button-control"
                  onClick={handleDelete(id)}
                  type="button"
                >
                  x
                </button>
              </div>
            </Row>
          </Col>
        </Row>
      ))}
    </div>
  );
};

export const TasksControl = ({ setFilterTask, deleteIsDone, activeTasksCount }) => {
  const [activeButton, setActiveButton] = useState('all');
  const handlerClick = (e) => {
    setFilterTask(e.target.name);
    setActiveButton(e.target.name);
  };
  const itemsView = activeTasksCount === 1 ? `${activeTasksCount} item left` : `${activeTasksCount} items left`;
  return (
    <Row className="tasks-control justify-content-between align-items-center">
      <div className="col-sm-3">{itemsView}</div>
      <div className="col-sm-auto">
        <button
          className={activeButton === 'all' ? 'button-filter button-active' : 'button-filter'}
          name="all"
          onClick={handlerClick}
          data-testid="filter-non"
          type="button"
        >
          All
        </button>
      </div>
      <div className="col-sm-auto">
        <button
          className={activeButton === 'active' ? 'button-filter button-active' : 'button-filter'}
          name="active"
          onClick={handlerClick}
          data-testid="filter-active"
          type="button"
        >
          Active
        </button>
      </div>
      <div className="col-sm-auto">
        <button
          className={activeButton === 'completed' ? 'button-filter button-active' : 'button-filter'}
          name="completed"
          onClick={handlerClick}
          data-testid="filter-completed"
          type="button"
        >
          Completed
        </button>
      </div>
      <div className="col-sm-auto">
        <button
          className="button-filter"
          onClick={deleteIsDone}
          type="button"
        >
          Clear completed
        </button>
      </div>
    </Row>
  );
};

const App = () => {
  const [tasks, setTasks] = useState([]);
  const [filterTask, setFilterTask] = useState('all');

  const activeTasksCount = tasks.filter(({ isDone }) => !isDone).length;

  const addTask = (textTask) => {
    if (textTask.trim() === '') {
      return;
    }
    const id = `t${Date.now()}`;
    const newTask = { id, text: textTask, isDone: false };
    setTasks((prevTasks) => {
      const newTasks = [newTask, ...prevTasks];
      return newTasks;
    });
  };

  const changeTask = (idTask = null) => {
    const updatedTasks = tasks.map((item) => {
      if (idTask === null) {
        return activeTasksCount === 0 ? { ...item, isDone: false } : { ...item, isDone: true };
      }
      return item.id === idTask ? { ...item, isDone: !item.isDone } : item;
    });
    setTasks(updatedTasks);
  };

  const deleteTask = (idTask) => {
    const updatedTasks = tasks.filter((item) => item.id !== idTask);
    setTasks(updatedTasks);
  };

  const deleteIsDone = () => {
    const updatedTasks = tasks.filter(({ isDone }) => !isDone);
    setTasks(updatedTasks);
  };

  return (
    <div className="container-main container">
      <header>
        <Row>
          <h1 className="header-todo">todos</h1>
        </Row>
      </header>
      <main>
        <div className="container-todo">
          <Container>
            <FormTask addTask={addTask} changeTask={changeTask} />
            <Tasks
              allTasks={tasks}
              changeTask={changeTask}
              deleteTask={deleteTask}
              filterTask={filterTask}
            />
            <TasksControl
              setFilterTask={setFilterTask}
              deleteIsDone={deleteIsDone}
              activeTasksCount={activeTasksCount}
            />
          </Container>
        </div>
      </main>
    </div>
  );
};

export default App;
