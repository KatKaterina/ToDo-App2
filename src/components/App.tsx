import React, { useState } from 'react';
import { Row, Container } from 'react-bootstrap';
import FormTask from './FormTask';
import Tasks from './Tasks';
import TasksControl from './TasksControl';

const App: React.FC = () => {
  const [filterTask, setFilterTask] = useState<string>('all');
  return (
    <div className="container-main container">
      <header>
        <Row>
          <h1 className="header-todo">todos</h1>
        </Row>
      </header>
      <main>
        <div className="container-todo">
          <Container>
            <FormTask />
            <Tasks filterTask={filterTask} />
            <TasksControl setFilterTask={setFilterTask} />
          </Container>
        </div>
      </main>
    </div>
  );
};

export default App;
