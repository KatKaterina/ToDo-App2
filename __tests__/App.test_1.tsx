import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect.js';
import userEvent from '@testing-library/user-event';
import React from 'react';

import App from '../src/components/App';

describe('App component', () => {
  it('App renders', () => {
    render(<App />);

    expect(screen.getByRole('form')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('What needs to be done?')).toBeInTheDocument();
  });

  it('App snapshot', () => {
    const todos = render(<App />);

    expect(todos).toMatchSnapshot();
  });
});

describe('Add task', () => {
  it('Add task', async () => {
    render(<App />);

    expect(screen.getByTestId('input-task')).toBeInTheDocument();
    expect(screen.getByTestId('input-task')).not.toHaveValue('New task');

    await userEvent.type(screen.getByTestId('input-task'), 'New task');
    expect(screen.getByTestId('input-task')).toHaveValue('New task');

    await userEvent.click(screen.getByText('+'));

    expect(screen.getByTestId('input-task')).not.toHaveValue('New task');
    expect(screen.getByText('New task')).toBeInTheDocument();
    expect(screen.getByText('New task')).toHaveClass('active-task');
  });
});
