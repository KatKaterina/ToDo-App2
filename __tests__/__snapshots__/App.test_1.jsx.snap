// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component App snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container-main container"
      >
        <header>
          <div
            class="row"
          >
            <h1
              class="header-todo"
            >
              todos
            </h1>
          </div>
        </header>
        <main>
          <div
            class="container-todo"
          >
            <div
              class="container"
            >
              <div
                class="new-task row"
              >
                <div
                  class="col-sm-2"
                >
                  <button
                    class="button-control"
                    type="button"
                  >
                    ✓
                  </button>
                </div>
                <div
                  class="col"
                >
                  <form
                    class=""
                    role="form"
                  >
                    <div
                      class="row"
                    >
                      <div
                        class="col"
                      >
                        <input
                          class="input-task"
                          data-testid="input-task"
                          placeholder="What needs to be done?"
                          type="text"
                          value=""
                        />
                      </div>
                      <div
                        class="col-sm-2"
                      >
                        <button
                          class="button-control"
                          type="submit"
                        >
                          +
                        </button>
                      </div>
                    </div>
                  </form>
                </div>
              </div>
              <div />
              <div
                class="tasks-control justify-content-between align-items-center row"
              >
                <div
                  class="col-sm-3"
                >
                  0 items left
                </div>
                <div
                  class="col-sm-auto"
                >
                  <button
                    class="button-filter button-active"
                    data-testid="filter-non"
                    name="all"
                    type="button"
                  >
                    All
                  </button>
                </div>
                <div
                  class="col-sm-auto"
                >
                  <button
                    class="button-filter"
                    data-testid="filter-active"
                    name="active"
                    type="button"
                  >
                    Active
                  </button>
                </div>
                <div
                  class="col-sm-auto"
                >
                  <button
                    class="button-filter"
                    data-testid="filter-completed"
                    name="completed"
                    type="button"
                  >
                    Completed
                  </button>
                </div>
                <div
                  class="col-sm-auto"
                >
                  <button
                    class="button-filter"
                    type="button"
                  >
                    Clear completed
                  </button>
                </div>
              </div>
            </div>
          </div>
        </main>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container-main container"
    >
      <header>
        <div
          class="row"
        >
          <h1
            class="header-todo"
          >
            todos
          </h1>
        </div>
      </header>
      <main>
        <div
          class="container-todo"
        >
          <div
            class="container"
          >
            <div
              class="new-task row"
            >
              <div
                class="col-sm-2"
              >
                <button
                  class="button-control"
                  type="button"
                >
                  ✓
                </button>
              </div>
              <div
                class="col"
              >
                <form
                  class=""
                  role="form"
                >
                  <div
                    class="row"
                  >
                    <div
                      class="col"
                    >
                      <input
                        class="input-task"
                        data-testid="input-task"
                        placeholder="What needs to be done?"
                        type="text"
                        value=""
                      />
                    </div>
                    <div
                      class="col-sm-2"
                    >
                      <button
                        class="button-control"
                        type="submit"
                      >
                        +
                      </button>
                    </div>
                  </div>
                </form>
              </div>
            </div>
            <div />
            <div
              class="tasks-control justify-content-between align-items-center row"
            >
              <div
                class="col-sm-3"
              >
                0 items left
              </div>
              <div
                class="col-sm-auto"
              >
                <button
                  class="button-filter button-active"
                  data-testid="filter-non"
                  name="all"
                  type="button"
                >
                  All
                </button>
              </div>
              <div
                class="col-sm-auto"
              >
                <button
                  class="button-filter"
                  data-testid="filter-active"
                  name="active"
                  type="button"
                >
                  Active
                </button>
              </div>
              <div
                class="col-sm-auto"
              >
                <button
                  class="button-filter"
                  data-testid="filter-completed"
                  name="completed"
                  type="button"
                >
                  Completed
                </button>
              </div>
              <div
                class="col-sm-auto"
              >
                <button
                  class="button-filter"
                  type="button"
                >
                  Clear completed
                </button>
              </div>
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
